src/algebra/category/FinVect.lean

```diff
@@ -0,0 +1,194 @@
+def FinVect := { V : Module.{u} K // finite_dimensional K V.carrier }
+
+namespace FinVect
+
+instance category : category (FinVect K) := by unfold FinVect; apply_instance
+
+instance finite_dimensional (V : FinVect K): finite_dimensional K V.val := V.prop
+
+instance has_sort_coe : has_coe_to_sort (FinVect K) := ⟨_, λ V, V.val⟩
+
+instance has_coe_to_fn (V W : FinVect K) : has_coe_to_fun (V ⟶ W) :=
+  ⟨λ _, V.val → W.val, λ f, f.to_fun⟩
+
+instance : inhabited (FinVect K) := ⟨⟨Module.of K K, finite_dimensional.finite_dimensional_self K⟩⟩
+
+/- This should go to `linear_algebra` after `linear_algebra` is cleaned up.
+  Right now it cannot go to either `linear_algebra.tensor_product` or
+  `linear_algebra.finite_dimension` because of dependencies. -/
+instance finite_dimensional_tensor_product (V V₂ : Type*) [add_comm_group V]
+  [module K V] [add_comm_group V₂] [module K V₂]
+  [finite_dimensional K V] [finite_dimensional K V₂] :
+  finite_dimensional K (tensor_product K V V₂) :=
+finite_dimensional.of_fintype_basis
+  (finsupp.basis.tensor_product (basis.of_vector_space K V) (basis.of_vector_space K V₂))
+
+instance monoidal_category : monoidal_category (FinVect K) :=
+@monoidal_category.full_monoidal_subcategory _ _ Module.monoidal_category
+ (λ (V : Module K), finite_dimensional K V)
+ (by { exact finite_dimensional.finite_dimensional_self K})
+ (λ X Y hX hY, @FinVect.finite_dimensional_tensor_product K _ X Y _ _ _ _ hX hY)
+
+-- This should go to `linear_algebra`.
+instance finite_dimensional_dual (V : Type*) [add_comm_group V] [module K V]
+  [finite_dimensional K V] :
+  finite_dimensional K (module.dual K V) :=
+finite_dimensional.of_fintype_basis (basis.dual_basis (basis.of_vector_space K V))
+
+variables (V : FinVect K)
+
+/-- The dual module is the dual in the rigid monoidal category `FinVect K`. -/
+def FinVect_dual : FinVect K :=
+⟨Module.of K (module.dual K V.val), FinVect.finite_dimensional_dual K V.val⟩
+
+open category_theory.monoidal_category
+
+/-- The coevaluation map is defined in `linear_algebra.coevaluation`. -/
+def FinVect_coevaluation : 𝟙_ (FinVect K) ⟶ V ⊗ (FinVect_dual K V) :=
+by { haveI := V.prop, change _ →ₗ[K] _, apply coevaluation K V.val }
+
+lemma FinVect_coevaluation_apply :
+ (FinVect_coevaluation K V) (1 : K) =
+   let bV := basis.of_vector_space K V.val in
+   ∑ (i : basis.of_vector_space_index K V.val), bV i ⊗ₜ[K] bV.coord i :=
+begin
+  simp only [FinVect_coevaluation, coevaluation, id],
+  rw [(basis.singleton unit K).constr_apply_fintype K],
+  simp only [fintype.univ_punit, finset.sum_const, one_smul, basis.singleton_repr,
+   basis.equiv_fun_apply,basis.coe_of_vector_space, one_nsmul, finset.card_singleton],
+end
+
+/-- The evaluation morphism is given by the contraction map. -/
+def FinVect_evaluation : (FinVect_dual K V) ⊗ V ⟶ 𝟙_ (FinVect K) :=
+(contract_left K V.val : _  →ₗ[K] K)
+
+lemma FinVect_evaluation_apply (f : (FinVect_dual K V).val) (x : V.val) :
```

Comment:
If you use the `@[derive ...]` trick mentioned above, it seems you can avoid needing the first `.val` here.